### ISSUE TITLE
using Lektor for LaTeX documents - possible or crazy?

### ISSUE THREAD

**Issue by bbbart** [lektor/lektor#1077]
I have a few Lektor websites running and really enjoy working with it. One of the things I love is that the CMS is really flexible and easy to set up and pretty rigid and simple for non-technical users to use.

Now, I was wondering if the lektor CMS/build system could be hacked to generate .tex files instead of .html (that, I believe, should be easy enough) and then to 'build' them into PDFs (for example).

Would it be a lot of work to get this workflow done with Lektor? Or is this a terrible idea? :-)
Or, does something what I am looking for already exist elsewhere, perhaps

Thanks for your insights!

**Comment by nixjdm** [comment]
I haven't done any LaTeX in a looong while, so my understanding of how to work with latex is pretty poor right now, maybe this is helpful. You could write a plugin to do basically anything you want in Python. I don't think I quite understand your use case though.

[Mistune's v3 release candidate has a built-in math plugin](https://github.com/lepture/mistune/blob/fc79e906449f1c03550d6329841be3fcc9e99a69/mistune/plugins/math.py), so you could I think use that to embed some latex in markdown, to generate html with math in it. You could also write a plugin to create a new content type so an entire field in a model could be pure latex, and again use that to produce html, or do anything you like with that latex with another plugin hook, like rendering pdfs.

I don't understand what you'd have as a source to then generate .tex files though. Are you hoping to use Lektor basically as an IDE to write .tex files? That seems backwards to me, but again, I'm fairly ignorant here. I'd have guessed you wanted to supply .tex files and render them into an html page for a website. I'm curious what you have in mind.

**Comment by bbbart** [comment]
Thanks for spending some thoughts on this, @nixjdm. I think 'using Lektor as a LaTeX-IDE' is probably the best way to put it. The workflow I imagine is:

 - I define a document model
 - I create a .tex template in jinja2
 - others use the Lektor CMS to create content
 - instead of publishing an HTML website, they download a LaTeX-compiled PDF document

I just really like the highly structured data modelling combined with the flexible templating of Lektor and I think this can be useful for LaTeX documents as well as HTML websites.

Perhaps the solution is simple:
 - I create .tex templates instead of .html
 - I create a post-processing plugin that runs LaTeX over the resulting files

I reckon that should be feasible, no (I have no experience with writing plugins for Lektor)?

**Comment by nixjdm** [comment]
That's all possible, and if you have some proficiency in Python you can do it.

For creating the `.tex` files, there's probably a few methods you could choose from. Lektor is really expecting to find `.html` templates, and render them as `.html` files. Changing the normal build process to look for a `.tex` template might be tough, but it's probably possible. I don't think I've seen that before. If you can make it easy and general, a plugin example or PRs to this repo might be a good idea.

Another approach might be to still have the template called a `.html` file, even though it isn't really, just so Lektor picks it up more easily, and then just control the built artifacts. There aren't many plugins that create custom artifacts, but there's a couple. You could draw some inspiration from these:

- https://www.getlektor.com/plugins/lektor-image-resize/
- https://www.getlektor.com/plugins/lektor-thumbnail-generator/

Yours would have some clear differences, but I'm sure `.tex` files could be generated. And then generating pdfs should be straight forward once you have the tex files.

However you do it, please share the result :) This seems like a novel example to learn from, if nothing else. It could join https://www.getlektor.com/plugins/ too if you think others might use this.

---

Because I don't fully understand the use-case, I wonder if Lektor is the appropriate tool. Jinja is fairly easy to call from Python, so perhaps a script to ingest templates, data stored however you like, and a tex-to-pdf renderer is the simplest approach. That could even tack on to a [after-build-all](https://www.getlektor.com/docs/api/plugins/events/after-build-all/) event hook, and not really leverage Lektor, so much as build artifacts alongside Lektor. Maybe you want to serve the pdfs in a website, but don't really need modeling for tex files. You have many options depending on your use.

**Comment by bbbart** [comment]
Thanks again. I just hacked together a quick plugin as a proof of concept. And it works!

For the templates, using the .tex extension is no problem, as long as I mention the full template file in the appropriate `contents.lr` file. The generated LaTeX sources indeed end on .html as per Lektor default, but `latexmk` doesn't care about that. :-)

I indeed hooked the after-build-all event and then `subprocess.Popen`-ed `latexmk` on the result, giving me a nice PDF.

Why I like to use Lektor for this (admitted, what follows will seem like another hack) is that I can host a permanently running `lektor server` behind an nginx reverse proxy on the network to allow less tech savvy people to generate beautiful documents using highly structured Lektor models (to avoid LaTeX compilation errors which make all sane people run away in severe anguish).

I'll clean up my hack and post it somewhere as a proof of concept. Of course, if deemed useful, I'd be more than happy to publish it for real too.

**Comment by nixjdm** [comment]
Great! As you seem to be on your development way, I'll close this issue. Please feel free to comment with anything you come up with though, or open PRs. Thanks!

**Comment by bbbart** [comment]
will do, thanks!

**Comment by bbbart** [comment]
So, meanwhile I've had a look into the plugin system and I tried out a couple of things. Using the `on_after_build_all` hook, I can invoke `latexmk` on rendered templates, much like https://github.com/chaos-bodensee/lektor-html-pretify/blob/main/lektor_html_pretify.py is doing.

However, immediately after, Lektor removes all my PDF (and other) files.

So, I went on to try and register them as artifacts to be retained by wrapping the `latexmk` call into a `BuildProgram`. However, that seems like a dead end, since I don't have access to the rendered templates at that point (or do I?). 

Some questions I'd like to discuss are:
 - should/can I manually invoke Jinja at this point? 
 - should/can I simply register the pdf's as artifacts to be retained here and combine this with the `on_after_build_all` hook?
 - would it be possible to, after compiling the LaTeX, replace the resulting html files with fixed templates embedding the resulting PDFs?

I'm coming to the conclusion that I seem to be trying to get Lektor to fit a use case it was not designed for, but I don't feel like quitting yet. :-)

Any comments on the above here?
